### PR TITLE
Add to_h Method to Telemetry V1 Classes

### PR DESCRIPTION
**Pull request by jennchenn**
**What does this PR do?**

This PR adds a `to_h` method to each of the new Telemetry V1 objects. This PR also removes the `Profiler` and `AppSec` classes and implements them as simple `Hash` objects instead.

**Motivation**

These methods will be used to serialize the telemetry payloads before sending telemetry events to the API.

**Additional Notes**

**How to test the change?**

Unit tests were added to each of the classes to check the `to_h` method returns the expected hash

### CODE REVIEW

File: spec/datadog/core/telemetry/v1/app_started_spec.rb [hunk 1]
```diff
@@ -122,4 +122,39 @@
...
+    context 'when attributes are all defined' do
+      let(:additional_payload) { { 'tracing.enabled' => true, 'profiling.enabled' => false } }
+      let(:configuration) { { DD_AGENT_HOST: 'localhost', DD_TRACE_SAMPLE_RATE: '1' } }
+      let(:dependencies) { [Datadog::Core::Telemetry::V1::Dependency.new(name: 'pg')] }
+      let(:integrations) { [Datadog::Core::Telemetry::V1::Integration.new(name: 'pg', enabled: true)] }
+
+      before do
+        allow(dependencies[0]).to receive(:to_h).and_return({ name: 'pg' })
+        allow(integrations[0]).to receive(:to_h).and_return({ name: 'pg', enabled: true })
+      end
```
Comment (ivoanjo) [inline spec/datadog/core/telemetry/v1/app_started_spec.rb:146]: Is this still needed? It looks to me that both of these get `#to_h` methods in this PR, so is there any difference between the actual implementations and the mock responses we're setting up here?

(Also, this happens on other files, so if you do change something based on my feedback, do a quick pass over all the files in the PR and also update other places where this is being used)

File: spec/datadog/core/telemetry/v1/app_started_spec.rb [hunk 2]
```diff
@@ -122,4 +122,39 @@
...
+      let(:additional_payload) { nil }
+      let(:configuration) { nil }
+      let(:dependencies) { nil }
+      let(:integrations) { nil }
+      it { is_expected.to eq({}) }
+    end
+
+    context 'when attributes are all defined' do
+      let(:additional_payload) { { 'tracing.enabled' => true, 'profiling.enabled' => false } }
+      let(:configuration) { { DD_AGENT_HOST: 'localhost', DD_TRACE_SAMPLE_RATE: '1' } }
```
Comment (ivoanjo) [inline spec/datadog/core/telemetry/v1/app_started_spec.rb:139]: These inputs don't seem very... realistic? E.g. according to the documentation in `#initialize` and other tests, `additional_payload` and `configuration` are expected to be both arrays (here we're just using hashes) and instances of `Telemetry::V1::Configuration`.

I actually think we may be hiding a bug here with these tests, since we're using `map_hash` in the code for these two fields and it seems like it should be `map_array`?

File: lib/datadog/core/telemetry/v1/app_started.rb [hunk 3]
```diff
@@ -22,6 +22,27 @@ def initialize(additional_payload: nil, configuration: nil, dependencies: nil, i
...
           end
+
+          def to_h
+            {}.tap do |app_started_hash|
+              app_started_hash[:additional_payload] = map_hash(@additional_payload) if @additional_payload
+              app_started_hash[:configuration] = map_hash(@configuration) if @configuration
+              app_started_hash[:dependencies] = map_array(@dependencies) if @dependencies
+              app_started_hash[:integrations] = map_array(@integrations) if @integrations
+            end
+          end
```
Comment (ivoanjo) [inline lib/datadog/core/telemetry/v1/app_started.rb:33]: Suggestion: according to #initialize, the fields stored by this class are arrays (or nil).

Consider storing them always as arrays, even if empty. I think this would simplify dealing with them overall, for instance I think the `#to_h` would be simplified as well.

Also Ruby tip: the `Kernel#Array()` function (yes, it's a function named `#Array`, it's not the class `Array`, can be used to very easily transform inputs, even when nil, into an Array:

```ruby
a = Array([1, 2])
# => [1, 2]
> b = Array(nil)
# => []
```

File: spec/datadog/core/telemetry/v1/application_spec.rb [hunk 4]
```diff
@@ -107,4 +107,43 @@
...
+          language_version: language_version,
+          products: { profiling: { version: '1.0' }, appsec: { version: '1.0' } },
+          runtime_name: runtime_name,
+          runtime_patches: runtime_patches,
+          runtime_version: runtime_version,
+          service_name: service_name,
+          service_version: service_version,
+          tracer_version: tracer_version }
+      )
+    end
```
Comment (ivoanjo) [inline spec/datadog/core/telemetry/v1/application_spec.rb:147]: Minor: You can actually omit the `{` / `}`here (like you did for other tests, actually).

(Note: This also applies to other files in this PR)

File: lib/datadog/core/telemetry/v1/product.rb [hunk 5]
```diff
@@ -14,6 +14,13 @@ def initialize(appsec: nil, profiler: nil)
             @appsec = appsec
             @profiler = profiler
           end
+
+          def to_h
+            {
+              appsec: @appsec.to_h,
+              profiler: @profiler.to_h
+            }
+          end
```
Comment (ivoanjo) [inline lib/datadog/core/telemetry/v1/product.rb:23]: I haven't been following along, but is there any doc for this telemetry format? Why are appsec and profiler inside a "producs object", but tracing isn't?

File: lib/datadog/core/telemetry/v1/profiler.rb [hunk 6]
```diff
@@ -14,6 +14,12 @@ def initialize(version:)
 
             @version = version
           end
+
+          def to_h
+            {
+              version: @version
+            }
+          end
         end
```
Comment (ivoanjo) [inline lib/datadog/core/telemetry/v1/profiler.rb:23]: I don't quite understand why is this needed? 🤔 

There's a single dd-trace-rb library, so this version will always be the same as the library version.

File: spec/datadog/core/telemetry/v1/telemetry_request_spec.rb [hunk 7]
```diff
@@ -164,4 +164,52 @@
...
+      Datadog::Core::Telemetry::V1::AppStarted.new(
+        integrations: [Datadog::Core::Telemetry::V1::Integration.new(
+          name: 'pg', enabled: true
+        )]
+      )
+    end
+    let(:request_type) { 'app-started' }
+    let(:runtime_id) { '20338dfd-f700-0d470f054ae8' }
+    let(:seq_id) { 42 }
+    let(:session_id) { '20338dfd-f700-0d470f054ae8' }
```
Comment (ivoanjo) [inline spec/datadog/core/telemetry/v1/telemetry_request_spec.rb:188]: Not that it matters a lot for this test, but these are documented as V4 UUIDs in the code, and these two examples are not V4 UUIDs which I find a bit confusing

My suggestion is, if we're using dummy data, we could using something like `dummy_runtime_id` rather than something that _looks like_ real data, but isn't. 😄

**Review by ivoanjo — COMMENTED** [review]
Left a few notes! 😄

File: spec/datadog/core/telemetry/v1/app_started_spec.rb (hunk 2 above)
Comment (jennchenn) [reply spec/datadog/core/telemetry/v1/app_started_spec.rb:139]: Based on https://github.com/DataDog/dd-trace-rb/pull/2102#discussion_r906509004, it was decided to change `additional_payload` and `configuration` to simple hashes. This change is reflected in `collector.rb` but not anywhere else. I was going to modify the rest of the code to reflect this in a separate PR, but I can make those changes in this one instead!

File: lib/datadog/core/telemetry/v1/product.rb (hunk 5 above)
Comment (jennchenn) [reply lib/datadog/core/telemetry/v1/product.rb:23]: I got this from: https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v1/SchemaDocumentation/Schemas/products.md

They seem to only expect `appsec` and `profiler` at the moment

File: spec/datadog/core/telemetry/v1/app_started_spec.rb (hunk 2 above)
Comment (ivoanjo) [reply spec/datadog/core/telemetry/v1/app_started_spec.rb:139]: Ah, got it. I think it's better to do it an separate PR, since this one is reasonably scoped and in a good shape :)

File: lib/datadog/core/telemetry/v1/product.rb (hunk 5 above)
Comment (ivoanjo) [reply lib/datadog/core/telemetry/v1/product.rb:23]: Hmmm... I guess they built it to be more generic and handle languages where appsec/profiling is separate, although it seems a quite biased model -- tracer is definitely the preferred child >_>

I'll bring it up with the team, but yeah I guess let's live with this for now

File: lib/datadog/core/telemetry/v1/profiler.rb (hunk 6 above)
Comment (jennchenn) [reply lib/datadog/core/telemetry/v1/profiler.rb:23]: The telemetry API expects the appsec/profiler version to be passed in in the products object: https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v1/SchemaDocumentation/Schemas/profiler.md

However since it's just one field, do you think it'd be better to remove the `AppSec` and `Profiler` classes and just use a regular Hash instead?

File: spec/datadog/core/telemetry/v1/telemetry_request_spec.rb [hunk 8]
```diff
@@ -38,9 +38,9 @@
     )
   end
   let(:request_type) { 'app-started' }
-  let(:runtime_id) { '20338dfd-f700-0d470f054ae8' }
+  let(:dummy_runtime_id) { '20338dfd-f700-0d470f054ae8' }
   let(:seq_id) { 42 }
-  let(:session_id) { '20338dfd-f700-0d470f054ae8' }
+  let(:dummy_session_id) { '20338dfd-f700-0d470f054ae8' }
   let(:tracer_time) { 1654805621 }
```
Comment (ivoanjo) [inline spec/datadog/core/telemetry/v1/telemetry_request_spec.rb:44]: Ah, apologies for not being clearer on my previous comment. I actually was talking about the value of the let, e.g.

```ruby
let(:runtime_id) { 'dummy_runtime_id' }
```

...not actually the name. (But it's also ok to use `let(:dummy_runtime_id) { 'dummy_runtime_id') `.

My concern is that I don't like using that data that looks real but isn't, because it can leave people thinking -- wait, is there another format other than UUID v4 that we need to support?

**Review by marcotc — APPROVED** [review]
👍 pending Ivo's comments.

File: spec/datadog/core/telemetry/v1/telemetry_request_spec.rb (hunk 8 above)
Comment (jennchenn) [reply spec/datadog/core/telemetry/v1/telemetry_request_spec.rb]: This makes a lot of sense - changed!

File: lib/datadog/core/telemetry/collector.rb [hunk 9]
```diff
@@ -179,17 +177,15 @@ def products
         end
 
         def profiler
-          version = profiler_version
-          Telemetry::V1::Profiler.new(version: version) if version
+          { version: profiler_version } if profiler_version
         end
```
[the remark this answers is not in the document]
Comment (jennchenn) [reply lib/datadog/core/telemetry/collector.rb:181]: Since we capture `profiling.enabled` in the `additional_payload`, I will change this to always send the profiler version in a future PR (and do the same for appsec)
